homework/index.js

```diff
@@ -29,18 +33,218 @@
+    selectList.addEventListener('change', () => {
+      selectList.parentElement.nextElementSibling.innerHTML = '';
+      cb(repositoriesData[selectList.value], repositoriesData[selectList.value].contributors_url);
+    });
+  };
+  // fetch repositories and contributions data as a callback hell so that once repositories data is fetched first, then the contribution data will be fetched.
+  const fetchRepositoriesAndContributions = url => {
+    fetchJSON(url, (errRepository, repositoriesData) => {
+      if (errRepository) {
+        createAndAppend('div', root, {
+          text: errRepository.message,
+          class: 'alert alert-danger display-3 text-center',
+        });
         return;
       }
-      createAndAppend('pre', root, { text: JSON.stringify(repositories, null, 2) });
+      appendRepositoriesToSelect(repositoriesData, (repositoryObj, contributionsURL) => {
+        createRepositoryLayout(repositoryObj);
+        fetchJSON(contributionsURL, (errContributions, contributionsObj) => {
+          if (errContributions) {
+            createAndAppend('div', root, {
+              text: `${errContributions.message}... The data cannot be fetched`,
+              class: 'alert alert-danger display-3 text-center',
+            });
+            return;
+          }
+          createContributionsLayout(contributionsObj);
+        });
+      });
+    });
+  };
+  // create page layout under the root div
+  const buildWebPage = url => {
+    const header = createAndAppend('header', root, {
+      class: 'container jumbotron mt-3 mb-3 px-5',
     });
+    createAndAppend('main', root, { id: 'main', class: 'container align-items-start' });
+    const headerDiv = createAndAppend('div', header, {
+      class: 'header-group d-flex flex-row justify-content-between mb-5 px-5',
+    });
+    createAndAppend('h2', headerDiv, {
+      text: 'Hack Your Future Repositories',
+      class: 'display-4 d-flex align-items-center mx-auto col-sm header-title',
+    });
+    const hyfLink = createAndAppend('a', headerDiv, {
+      href: 'https://www.hackyourfuture.net/',
+      target: '_blank',
+    });
+    createAndAppend('img', hyfLink, {
+      class: 'img-fluid rounded img-thumbnail mx-auto',
+      src: './hyf.png',
+      alt: 'hack your future thumbnail',
+    });
+    const selectList = createAndAppend('select', header, {
+      id: 'repo-select',
+      class: 'form-control',
+      'aria-label': 'Hack Your Future Repositories Selection',
+    });
+    // Create Select Box and its options
+    createAndAppend('option', selectList, {
+      value: ' ',
+      disabled: '',
+      selected: '',
+      text: 'Select a repository',
+      class: 'text-muted options',
+    });
+    fetchRepositoriesAndContributions(url);
+  };
+  // main() ==> buildWebPage() ==> fetchRepositoriesAndContributions() ==> appendRepositoriesToSelect() ==>  createRepositoryLayout() && createContributionsLayout()
+  function main(url) {
+    buildWebPage(url);
   }
```

Comment:
I would move your call to `fetchRepositoriesAndContributions` to `main()`. In that case `buildWebPage()` no longer needs a `url` parameter:

```js
function main(url) {
  buildWebPage();
  fetchRepositoriesAndContributions(url);
}
```

Reply:
I have done it.